src/toil/spark/spark_cluster.py

```diff
@@ -0,0 +1,439 @@
+    if docker_parameters:
+        base_docker_call = base_docker_call + docker_parameters
+
+    _log.warn("Calling docker with %s." % " ".join(base_docker_call + [tool] + tool_parameters))
+    print >> _fp, "Calling docker with %s." % " ".join(base_docker_call + [tool] + tool_parameters)
+
+    try:
+        if outfile:
+            subprocess.check_call(base_docker_call + [tool] + tool_parameters, stdout=outfile)
+        else:
+            if check_output:
+                return subprocess.check_output(base_docker_call + [tool] + tool_parameters)
+            else:
+                subprocess.check_call(base_docker_call + [tool] + tool_parameters)
+
+    except subprocess.CalledProcessError:
+        raise RuntimeError('docker command returned a non-zero exit status. Check error logs.')
+    except OSError:
+        raise RuntimeError('docker not found on system. Install on all nodes.')
+
+
+def _checkContainerStatus(sparkContainerID,
+                          hdfsContainerID,
+                          sparkNoun='leader',
+                          hdfsNoun='namenode'):
+
+    containers = subprocess.check_output(["docker", "ps", "-q"])
+    print >> _fp, "Checking for %s/%s/%s/%s, got %r" % (sparkContainerID,
+                                                        hdfsContainerID,
+                                                        sparkNoun,
+                                                        hdfsNoun,
+                                                        containers)
+
+    # docker ps emits shortened versions of the hash
+    # these shortened hashes are 12 characters long
+    shortSpark = sparkContainerID[0:11]
+    shortHdfs = hdfsContainerID[0:11]
+
+    if ((sparkContainerID not in containers and
+         shortSpark not in containers) or
+        (hdfsContainerID not in containers and
+         shortHdfs not in containers)):
+        raise RuntimeError('Lost both Spark %s and HDFS %s.' % (sparkNoun, hdfsNoun))
+    elif sparkContainerID not in containers and shortSpark not in containers:
+        raise RuntimeError('Lost Spark %s. %r' % sparkNoun)
+    elif hdfsContainerID not in containers and shortHdfs not in containers:
+        raise RuntimeError('Lost HDFS %s. %r' % hdfsNoun)
+    else:
+        return True
+    
+
+class SparkService(Job.Service):
+    """
+    A Service job that spins up a Spark cluster that child jobs can then attach
+    to. If the job that spawns this job is run with `checkpoint = True`, then
+    this service will robustly restart the Spark cluster upon the loss of any
+    nodes in the cluster.
+    """
+
+    def __init__(self,
+                 sudo,
+                 memory,
+                 cores=None,
+                 overrideLeaderIP=None):
+        """
+        :param sudo: Whether this code should run docker containers with sudo.
+        :param memory: The amount of memory in Gibibytes that should be \
```

Comment:
Okay, I think we have a consistency issue here in how we specify resource requirements. We have been more flexible in resource specification, and for services we follow the convention as for jobs:

https://github.com/BD2KGenomics/toil/blob/master/src/toil/job.py#L64

See:

https://github.com/BD2KGenomics/toil/blob/master/src/toil/job.py#L890

This needs to be fixed for sure.